### PR TITLE
hpa update policy

### PR DESCRIPTION
**Pull request by minbaev**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:
Adding UpdatePolicy structure to HPA spec. This is needed to be able to run HPA in dry run mode

#### Which issue(s) this PR fixes:

Fixes #
https://github.com/kubernetes/enhancements/issues/3148 

#### Does this PR introduce a user-facing change?
None

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/3149/commits/ce63d3069caa7c9331d86f819fd7d1c939ef5d9d
```

### CODE REVIEW

**Comment by minbaev** [comment]
@josephburnett

File: pkg/apis/autoscaling/types.go [hunk 1]
```diff
@@ -101,8 +101,33 @@ type HorizontalPodAutoscalerSpec struct {
...
 
+// UpdateMode controls when autoscaler applies changes to the cluster.
+type UpdateMode string
+
+const (
+	// UpdateModeOff means that autoscaler never cretes/deletes pods.
+	// The HorizontalPodAutoscalerStatus.DesiredReplicas field is still populated.
+	// This can be used for a "dry run"
+	UpdateModeOff UpdateMode = "Off"
+	// UpdateModeAuto means that autoscaler creates/deletes pods based on the sclaing decision
```
Comment (josephburnett) [inline pkg/apis/autoscaling/types.go:127]: Nit: spelling "scaling"

File: staging/src/k8s.io/api/autoscaling/v2beta2/types.go [hunk 2]
```diff
@@ -81,8 +81,32 @@ type HorizontalPodAutoscalerSpec struct {
 	// If not set, the default HPAScalingRules for scale up and scale down are used.
 	// +optional
 	Behavior *HorizontalPodAutoscalerBehavior `json:"behavior,omitempty" protobuf:"bytes,5,opt,name=behavior"`
+
+	// UpdatePolicy describes how changes are applied to the cluster.
+	// If not specified, all values are set to default
+	UpdatePolicy *HorizontalPodAutoscalerUpdatePolicy `json:"updatePolicy,omitempty" protobuf:"bytes,6,opt,name=updatePolicy"`
```
Comment (josephburnett) [inline staging/src/k8s.io/api/autoscaling/v2beta2/types.go:87]: Please do not add new fields to the v2beta2 API as it is [deprecated](https://github.com/kubernetes/kubernetes/blob/393337dc84a90979f23f0cf11520bf1adb995a03/staging/src/k8s.io/api/autoscaling/v2beta2/types.go#L30).  You can just add this new field directly to the v2 (stable) API alone.

File: staging/src/k8s.io/api/autoscaling/v2/types.go [hunk 3]
```diff
@@ -9,7 +9,7 @@ You may obtain a copy of the License at
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.ß
```
Comment (josephburnett) [inline staging/src/k8s.io/api/autoscaling/v2/types.go:12]: Nit: remove "ß"

File: staging/src/k8s.io/api/autoscaling/v2/types.go [hunk 4]
```diff
@@ -79,8 +79,32 @@ type HorizontalPodAutoscalerSpec struct {
...
 
+// UpdateMode controls when autoscaler applies changes to the cluster.
+type UpdateMode string
+
+const (
+	// UpdateModeOff means that autoscaler never cretes/deletes pods.
+	// The HorizontalPodAutoscalerStatus.DesiredReplicas field is still populated.
+	// This can be used for a "dry run"
+	UpdateModeOff UpdateMode = "Off"
+	// UpdateModeAuto means that autoscaler creates/deletes pods based on the sclaing decision
```
Comment (josephburnett) [inline staging/src/k8s.io/api/autoscaling/v2/types.go:104]: Nit: spelling "scaling"

**Comment by atiratree** [comment]
until a KEP is resolved
/hold

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by dims** [comment]
If you still need this PR then please remove the merge commits (see bot comment above), if not, please close the PR

**Comment by soltysh** [comment]
Since we haven't heard from the author for this long, I'm closing this PR. Feel free to re-open this PR once you have the KEP accepted, just like @atiratree mentioned above. 
/close

**Pull request closed without merging**